server/src/route/mod.rs

```diff
@@ -5,14 +5,56 @@ use std::sync::Arc;
 use async_trait::async_trait;
 use ceresdbproto::storage::{Route, RouteRequest};
 
-use crate::error::Result;
-
 pub mod cluster_based;
 pub(crate) mod hash;
 pub mod rule_based;
 
 pub use cluster_based::ClusterBasedRouter;
 pub use rule_based::{RuleBasedRouter, RuleList};
+use snafu::{Backtrace, Snafu};
+
+#[derive(Snafu, Debug)]
+#[snafu(visibility(pub))]
+pub enum Error {
+    #[snafu(display(
+        "No route found for schema, schema:{}.\nBacktrace:\n{}",
+        schema,
+        backtrace
+    ))]
+    RouteNotFound {
+        schema: String,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display(
+        "No shard found for metric, schema:{}, table:{}.\nBacktrace:\n{}",
+        schema,
+        table,
+        backtrace
+    ))]
+    ShardNotFound {
+        schema: String,
+        table: String,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Failed to parse endpoint, endpoint:{}, err:{}", endpoint, source))]
+    ParseEndpoint {
+        endpoint: String,
+        source: Box<dyn std::error::Error + Sync + Send>,
+    },
+
+    #[snafu(display("Failure caused by others, msg:{}, err:{}", msg, source))]
+    OtherWithCause {
+        msg: String,
+        source: Box<dyn std::error::Error + Sync + Send>,
+    },
+
+    #[snafu(display("Failure caused by others, msg:{}.\nBacktrace:\n{}", msg, backtrace))]
+    OtherNoCause { msg: String, backtrace: Backtrace },
```

Comment:
I see `OtherNoCause` is unused now?

Reply:
It is used in `server/src/route/cluster_based.rs`.